msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/HttpRequest.java

```diff
@@ -87,4 +87,36 @@ private URL createUrlFromString(String stringUrl) {
 
         return url;
     }
+
+    // Adding this for unit tests to be able to compare expected and actual request.
+    @Override
```

Comment:
Use https://projectlombok.org/features/EqualsAndHashCode ?

That said, ideally we should not change product code like this just to make it testable? We will forget about this and add HttpRequsts to some dictionary / map and get wierd bugs.

It'd be better to have a helper in tests that helps with assertions and which can access httpMethod, headers, body etc.

Reply:
I did not find a way to add helpers to compare the request in tests since the request is framed in the flow in private methods. This is how the mocking is in tests `when(httpClientMock.send(eq(expectedRequest(source, resource)))).thenReturn(expectedResponse(200, ""));`

I have added the annotation, but not sure if there is a way to write a helper to verify the request.

Another option is to remove the validation for request.

@Avery-Dunn to help confirm if there is any other way.

Reply:
Since the URL, HTTP method, headers, and body are all fields in HTTPRequest, then Lombok's EqualsAndHashCode annotation should create an equals method that's looks like the return method you have here.

If the comparisons still aren't working, then it might be cleaner to just have this method in the relevant tests like @bgavrilMS said.

Reply:
Updated the HttpRequest to add Lombok's EqualsAndHashCode and it works.